File: Irrgarten-ruby/UI/text_ui.rb
# frozen_string_literal: true

require 'io/console'
require_relative '../Irrgarten/directions'
require_relative '../Irrgarten/game_state'


module UI

  class TextUI

    def read_char
      STDIN.echo = false
      STDIN.raw!

      input = STDIN.getc.chr
      if input == "\e"
        input << STDIN.read_nonblock(3) rescue nil
        input << STDIN.read_nonblock(2) rescue nil
      end
    ensure
      STDIN.echo = true
      STDIN.cooked!

      return input
    end

    def next_move
      print "Where? "
      got_input = false
      while (!got_input)
        c = gets.chomp
        case c
        when "w"
          puts "UP ARROW"
          output = Irrgarten::Directions::UP
          got_input = true
        when "s"
          puts "DOWN ARROW"
          output = Irrgarten::Directions::DOWN
          got_input = true
        when "d"
          puts "RIGHT ARROW"
          output = Irrgarten::Directions::RIGHT
          got_input = true
        when "a"
          puts "LEFT ARROW"
          output = Irrgarten::Directions::LEFT
          got_input = true
        when "\u0003"
          puts "CONTROL-C"
          got_input = true
          exit(1)
        else
          #Error
        end
      end
      output
    end

    def show_game(get_game_state)
      puts "Representación del laberinto: \n"
      puts get_game_state.labyrinthv
      puts "Jugador actual: \n"
      puts (get_game_state.current_players).to_s
      puts "Representación de los monstruos: \n"
      puts get_game_state.monsters
      puts "Representación de los jugadores: \n"
      puts get_game_state.players
      puts "--------------------------------------"
      puts "Representación del log: \n"
      puts get_game_state.log

      puts "Representación del ganador: \n"
      puts get_game_state.winner

    end

  end # class

end # module

File: Irrgarten-Java/src/UI/TextUI.java

package UI;

import irrgarten.*;
//import irrgarten.Directions;
//import irrgarten.GameState;
import java.util.Scanner;


public class TextUI {
    
    private static Scanner in = new Scanner(System.in);
    
    private char readChar() {
        String s = in.nextLine();     
        return s.charAt(0);
    }
    

    public Directions nextMove() {
        System.out.print("Where? ");
        
        Directions direction = Directions.DOWN;
        boolean gotInput = false;
        
        while (!gotInput) {
            char c = readChar();
            switch(c) {
                case 'w':
                    System.out.print(" UP\n");
                    direction = Directions.UP;
                    gotInput = true;
                    break;
                case 's':
                    System.out.print(" DOWN\n");
                    direction = Directions.DOWN;
                    gotInput = true;
                    break;
                case 'd':
                    System.out.print("RIGHT\n");
                    direction = Directions.RIGHT;
                    gotInput = true;
                    break;
                case 'a':
                    System.out.print(" LEFT\n");
                    direction = Directions.LEFT;
                    gotInput = true;    
                    break;
            }
        }    
        return direction;
    }
    
    public void showGame(GameState gameState) {  
        System.out.println("Representación del laberinto: \n");
        System.out.println(gameState.getLabyrinthv());
        System.out.println("Jugador actual: \n");
        System.out.println(Integer.toString(gameState.getCurrentPlayer()));
        System.out.println(gameState.getPlayers());
        System.out.println("Representación del monstruo: \n");
        System.out.println(gameState.getMonsters());
        System.out.println("--------------------------------------");
        System.out.println("Representación del log: \n");
        System.out.println(gameState.getLog());
        System.out.println("Representación del ganador: \n");
        System.out.println(gameState.getWinner());
        
    }
    
}
